Port to V8

skttl opened the pull request:
**✨✨[If you want to try out the port, the latest build can be downloaded here](https://ci.appveyor.com/project/UMCO/umbraco-doc-type-grid-editor/build/artifacts)✨✨**

Here's a new PR with my work.

I got the UI to work, so you can select a doctype and load the properties of it, (or just load the properties if you are in "single doctype" mode).

There is some issues with the media picker (and probably other data types) though.

There is still work to do in regards to the backend code. I tried changing the output from IPublishedContent to IPublishedElement, but it's not generating the "Detached" element correctly (or not sending it correctly through)

I think this is where my abilities is lacking. Hopefully others can chip in, and get the port going.

bjarnef commented on src/Our.Umbraco.DocTypeGridEditor/Extensions/ContentTypeServiceExtensions.cs:
```diff
@@ -9,12 +9,15 @@ internal static class ContentTypeServiceExtensions
+                {
+                    using (var scope = Current.ScopeProvider.CreateScope())
+                        return scope.Database.ExecuteScalar<string>("SELECT [cmsContentType].[alias] FROM [cmsContentType] INNER JOIN [umbracoNode] ON [cmsContentType].[nodeId] = [umbracoNode].[id] WHERE [umbracoNode].[uniqueID] = @0", id);
+                });
```
@skttl I think this need to call `scope.Complete();` or add `autoComplete: true` in `CreateScope()`.
https://our.umbraco.com/forum/umbraco-8/96075-alternative-solution-to-use-applicationcontextcurrentdatabasecontextdatabase-in-umbraco-8

ArnoldV commented:
Just a thought... because of IPublishedElement, should DetachedElement still be used? Arn’t they the ‘same thing’?

bjarnef commented:
It would be great to use position `target` for `umb-overlay`, so it is consistent with core overlay to insert grid editors. Furthermore does the editor need to select `doc type grid editor` first as in v7 (not sure is this have been changed to list the defined editors merged with other grid editors).
https://github.com/umco/umbraco-doc-type-grid-editor/blob/36d35c243abb36e9ea91d3175a197fc25a1d9619/src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html#L15-L19

![image](https://user-images.githubusercontent.com/2919859/54267953-828b3a80-457a-11e9-9cf3-746cdb886aa9.png)

skttl commented:
@ArnoldV if we can use IPublishedElement, then I think we should!

skttl commented:
@bjarnef `Furthermore does the editor need to select doc type grid editor first as in v7 (not sure is this have been changed to list the defined editors merged with other grid editors).`

That depends on how you use DTGE. The default editor it installs, just lets you select every doctype (in the dialog). I think (hope) that most people don't use it that way, and are instead creating different editors for different doctypes. Then they would show up alongside other grid editors too.

bjarnef commented:
@skttl yes, and the default behaviour also require an extra click. I would be great it created a grid editor for each document type similar to LeBlender and when creating own custom grid editors which doesn't rely on third party packages.

skttl commented:
> 
> 
> @skttl yes, and the default behaviour also require an extra click. It would be great if it created a grid editor for each document type similar to LeBlender and when creating own custom grid editors which doesn't rely on third party packages.

But how would DTGE know which doctypes to create grid editors for?

You could say that DTGE should have a GUI, but I think that would be unnecessary. Especially if Grid2 is going to be a thing - then DTGE is obsolete.

bjarnef commented:
Yes, a GUI to manage grid editors might be part of the Grid v2. At the moment developers can create a website only from backoffice, but to manage grid editors you need to edit grid editors in Visual Studio, VS Code etc. or manage them like in LeBlender.  
Unfortunately LeBlender didn't support grid editors in plugin folders.
Not sure if Grid v2 is planned for Umbraco v8.1 or in another future release.

soreng commented:
> Unfortunately LeBlender didn't support grid editors in plugin folders.

yes and no, it shows them in the tree 
> https://github.com/Lecoati/LeBlender/blob/dev-v1/Src/Lecoati.LeBlender.Extension/Controllers/PropertyGridEditorController.cs#L36
But it can only write to the config file:
https://github.com/Lecoati/LeBlender/blob/dev-v1/Src/Lecoati.LeBlender.Extension/Controllers/HelperController.cs#L31

NathalieLarsson commented:
We have a working version of dtge (based on this PR) with a leblender inspired solution as gui on our Umbraco 8 installation. It got a bunch of custom stuff that's specific to our needs, but I can try to make a PR with the basic dtge stuff later this week if you want.?

rasmusfjord commented:
That would be awesome Søren :) Especially if it includes the right click
thing you build on the doctype, that is an awesome addition.

On Mon, Apr 1, 2019 at 1:14 PM Nathalie Festin <notifications@github.com>
wrote:

> We have a working version of dtge (based on this PR) with a leblender
> inspired solution as gui on our Umbraco 8 installation. It got a bunch of
> custom stuff that's specific to our needs, but I can try to make a PR with
> the basic dtge stuff later this week if you want.?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/umco/umbraco-doc-type-grid-editor/pull/137#issuecomment-478537765>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOfmy3zmicq4tZTwdJje8T1jtw2Vn8rks5vceoqgaJpZM4bnbgU>
> .
>


-- 
-----------------------------------------------
Rasmus Lykke Fjord
+45 42 40 76 20

skttl commented:
@nathalielarsson that would be super awesome!

0Neji commented:
Just wanted to know if any progress had been made on this one? I'd be interested in helping out to get this up and running while we wait for Grid v2!

skttl commented:
@NathalieLarsson have you had any chance to look at a PR yet?

skttl commented:
Just spent some time in the dream corner, trying to make this work. And I actually did. It doesn't work automatically with models builder, but I'll try to clean up and fix some other bugs, and then I'll push some updates 🍾

0Neji commented:
Awesome, thanks @skttl - I was planning to have a look at this next week myself so would be great if you could push your work, possibly even if you don't get ModelsBuilder running. More than happy to help out if need be, sounds like you might have it hand though!

skttl commented:
So I added the bits, that make it work as IPublishedElements. There is a bug in the backoffice, where the dialog doesn't close on save - haven't fixed that one yet.

You can use Models Builder, you have to do like this in your view (or your hijacked surfacecontroller)
```
@inherits UmbracoViewPage<IPublishedElement>
@{ 
    var myModel = new MyModel(Model);
}
```

The problem about ModelsBuilder is that the factory for using the model, is internal on the Current scope (good thing though), and I am not as experienced in using DI. It can be injected using IPublishedModelFactory.

skttl commented:
So, ModelsBuilder now works automatically, the preview also works, and the dialog works as expected when saving the content.

You can give it a spin - download an Umbraco package at https://ci.appveyor.com/project/UMCO/umbraco-doc-type-grid-editor/builds/24832157/artifacts

Current issues (there might be more):
- media picker doesn't work, the dialog for the picker is faded, and once you select a media and close the dialog, the DTGE dialog closes also.
- Default DtgeSurfaceController resolver is commented out, need to find out how to get that one working.

skttl commented:
I have now added the ability to route hijack using surface controllers (see current docs, as it works the same way).

Also added the ability the change the default surfacecontroller. Not sure it is documented, but it was possible using `DefaultDocTypeGridEditorSurfaceControllerResolver`. It is now done using an `IUserComposer` like you would use for all other composition tasks in v8.

I have added an extension method to `IComposition` for setting the `DefaultDocTypeGridEditorSurfaceControllerType` on DTGEs own (new) `Current` object, inspired by how you change the default RenderMvcController in v8.

Would love some feedback, as this is not something I am experienced with.

In the backoffice, I switched from using `umb-overlay` to using the new `editorService`. This brings the ability to use infinite editing, and it also fixed the bug regarding the media picker.

There is a few TODOs left from the original PR. Can anyone verify that these are all right? To me it seems like everything is working as it should.

- https://github.com/skttl/umbraco-doc-type-grid-editor/blob/v8upgrade/src/Our.Umbraco.DocTypeGridEditor/Models/DetachedPublishedProperty.cs#L28
- https://github.com/skttl/umbraco-doc-type-grid-editor/blob/v8upgrade/src/Our.Umbraco.DocTypeGridEditor/Models/UnpublishedContent.cs#L141
- https://github.com/skttl/umbraco-doc-type-grid-editor/blob/v8upgrade/src/Our.Umbraco.DocTypeGridEditor/Web/Controllers/DocTypeGridEditorApiController.cs#L133
- https://github.com/skttl/umbraco-doc-type-grid-editor/blob/v8upgrade/src/Our.Umbraco.DocTypeGridEditor/Web/Controllers/DocTypeGridEditorApiController.cs#L161

cleversolutions commented:
Thanks @skttl and all, your hard work is very much appreciated by the community.

ArnoldV commented:
Currently testing with this grid setup:

` {
      "name": "Grid Blocks",
      "alias": "docTypeGridEditor",
      "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
      "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
      "icon": "icon-hand-pointer-alt",
      "config": {
        "allowedDocTypes": [],
        "nameTemplate": "",
        "enablePreview": true,
        "viewPath": "/Views/Partials/Grid/Editors/",
        "previewViewPath": "",
        "previewCssFilePath": "",
        "previewJsFilePath": ""
      }
    }`

So now predefined doctypes:

When selecting it gives me this view:
![image](https://user-images.githubusercontent.com/4798968/58462823-f1684080-8132-11e9-819a-dc34b3581000.png)

That is as you currently expect it to be?

skttl commented:
Hi Arnold, that definitely doesn't look right. Maybe has to do something with using editorService instead of umb-overlay. I'll fix it up later - if you don't beat me to it :) You can add a PR to my fork on the v8upgrade if you want :)

ArnoldV commented:
Created two pull requests to your branch to get this view with filter:

![image](https://user-images.githubusercontent.com/4798968/58470020-182d7380-8141-11e9-8035-6569caf6a776.png)


Also open for discussing: should "GetContentTypes" only return PublishedElement content types (like how Nested Content Works) or not?

mattbrailsford commented:
@ArnoldV I'd say yes, you just want to pull back `PublishedElement` doc types back

mattbrailsford commented:
@ArnoldV just an FYI, in other packages (like stacked content) we've gone for more of the list approach for selecting things as this allows a description to be displayed as well

![image](https://user-images.githubusercontent.com/527305/58470403-853d0b00-8139-11e9-8efd-938a9f8ec9d6.png)

skttl commented:
@mattbrailsford @ArnoldV 

I think the grid of types could work fine if you have many (10+) doctypes to choose from. Maybe a config setting on the grid editor, to pick between grid/list view?

mattbrailsford commented:
@skttl personally I think it's harder to skim the grid compared to a list so I'd probably go with the list. But that's just my POV

bjarnef commented:
@skttl the list approach would also allow to use description from element types, but maybe it should be configurable. Often its a great help for editors with the description and e.g. when having multiple grid editors with similar names.

ArnoldV commented:
The grid approach is more in line with what an editor sees when adding elements in the grid.
Naming can help a lot there. I chose this view since it's a default Umbraco Element.

Most of our implementations we don't even show the choose item view, we create a 1 doctype per editor config so the editor view is shown directly. 

Does someone know if you can overrule the 'edit item' header the grid default gives you?

skttl commented:
@ArnoldV the dialog title is the title property on the overlayOptions.
src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Js/doctypegrideditor.controllers.js

skttl commented:
@mattbrailsford (and @leekelleher) it is your package, so I think it ultimately is your decision if we should go with grid, lists or both (config) for the doctype selection.

leekelleher commented:
@skttl Personally, I like the doctype list with descriptions ... BUT! DTGE currently does a dropdown/select to pick a doctype from, so in the spirit of _getting stuff done_, I'd say go with whatever you've currently done. Otherwise we'd fall into bike-shedding. 🙏

skttl commented:
I actually think it does the list (think I did the PR 😎), but I don't have strong opinions about it. I almost never use the doctype selector.

ArnoldV commented:
I can do a list view alternative quite easily probably later today, have some ideas... Still looking for a good example of styling something like that in the backoffice. Keep you updated ;)

bjarnef commented on src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.dialog.html:
```diff
@@ -1,35 +1,60 @@
+                    <ul class="umb-card-grid -three-in-row">
+                        <li ng-repeat="docType in docTypes | orderBy:'name' | filter:searchTerm"
+                            ng-click="selectDocType(docType.alias)">
+                            <a class="umb-card-grid-item" href="" title="{{ docType.name }}">
```
@skttl can we update `title` to `ng-attr-title` instead, which seems to be supported since Angular v1.1.4?
https://stackoverflow.com/a/18346925/1693918

It seems it also should be fixed in core
https://github.com/umbraco/Umbraco-CMS/blob/1815112ae7a3433ca829bd5537f4417fea883b38/src/Umbraco.Web.UI.Client/src/views/common/infiniteeditors/datatypepicker/datatypepicker.html#L50

bjarnef commented:
@skttl it doesn't seem to remember the selected link/node using "Url Picker".

![image](https://user-images.githubusercontent.com/2919859/58483864-4e7aeb00-8161-11e9-959e-5872e23d6aa8.png)

When re-opening the editor it is empty.

![image](https://user-images.githubusercontent.com/2919859/58483960-766a4e80-8161-11e9-86f6-af9f6b5ce34d.png)

ArnoldV commented:
WIP: 
![image](https://user-images.githubusercontent.com/4798968/58489633-eed60d00-816b-11e9-90cf-11a8ff12ba8d.png)

you can set the grid layout or the list layout via config setting.
only need to add the descriptions to the listview. Will do a PR tomorrow for this probably.

skttl commented:
@bjarnef should be fixed now, Nested Content and some other editors had the same problem. They were expecting a formSubmitted event before saving the value to their $scope.model

skttl commented:
I added a config setting (showDocTypeSelectAsGrid) to select between grid and list when selecting doctypes. @ArnoldV, I added markup from the create dialog on content, to make it visually align with the backoffice.

Also ported the blueprint stuff I made for 1.0.0, so now this one supports blueprints too.

And then I made the titlebar of the dialog respond to the dialog mode, so it says Select Content Type, Select Content Template, Edit Item etc.

skttl commented:
Oh, and I restricted the allowed doctypes to be element types only.

skttl commented:
![2019-05-28_21-20-16](https://user-images.githubusercontent.com/3726467/58505980-8c423880-818e-11e9-896b-aa405522c750.gif)

bjarnef commented:
@skttl maybe wrap the list of document types inside `<umb-box>` directive/component so it has the white background similar to the box wrapping the properties.

bjarnef commented:
The `MinUmbracoVersion` property should be updated to 8.0.0 so the NuGet package requires Umbraco v8.0.0 or newer.
https://github.com/skttl/umbraco-doc-type-grid-editor/blob/v8upgrade/build/package.proj#L21

You could update the copyright year to 2019 as well.
![image](https://user-images.githubusercontent.com/2919859/58507517-3b344380-8192-11e9-845c-28e3638de941.png)

ArnoldV commented:
@skttl @bjarnef in my example I used an umb-table. The styling then resembles the listview which is used for picking nodes in certain overlays as well.

skttl commented:
Thanks, much better :)
![billede](https://user-images.githubusercontent.com/3726467/58510276-a8e36e00-8198-11e9-867d-d6799b8b7d64.png)

skttl commented:
maybe needs some padding on the search though

bjarnef commented:
you should probably also add the inner `<umb-box-content>` https://github.com/umbraco/Umbraco-CMS/blob/853087a75044b814df458457dc9a1f778cc89749/src/Umbraco.Web.UI.Client/src/views/macros/views/settings.html#L13

bjarnef commented:
When inserting the DGTE but you haven't selected a document type, it just show the icon in the grid.

![image](https://user-images.githubusercontent.com/2919859/58535546-9431c500-81ee-11e9-87b0-eddc800f3a63.png)

compared to when a document type is selected

![image](https://user-images.githubusercontent.com/2919859/58535573-a7dd2b80-81ee-11e9-94ef-7703016dd11f.png)

ArnoldV commented:
@bjarnef maybe we can also make it bigger and more 'block like' while at it?

leekelleher commented:
There's already some CSS (commented out) that makes the previews nicer...
https://github.com/umco/umbraco-doc-type-grid-editor/blob/develop/src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Css/doctypegrideditor.css#L21-L43

(I can't recall the reason we commented it out at the time 🤷‍♂)

bjarnef commented:
@ArnoldV yes, something like this would be nice.

![image](https://user-images.githubusercontent.com/2919859/58537273-d9f08c80-81f2-11e9-8ebd-30c7b8034136.png)

ArnoldV commented:
@skttl would you like to do this or do you want me to spend a bit of time on it?

bjarnef commented:
It seems there are some issues when making a change and publish changes, the changes are updated in backoffice + in preview, but not on site.

I have tested it on Umbraco 8.0.0, so it might be fixed in one of the latest patch updates. Do you have this issue as well @skttl ?

![image](https://user-images.githubusercontent.com/2919859/58539512-ca277700-81f7-11e9-9d76-825e6b17eea2.png)

![image](https://user-images.githubusercontent.com/2919859/58539579-f04d1700-81f7-11e9-8498-e50b0b902521.png)

skttl commented:
Go ahead @ArnoldV :)

@bjarnef just tried 8.0.2, does the same. I think it has to do with DTGE caching the value. I'll look into that.

skttl commented:
The caching had been changed to RuntimeCache in https://github.com/umco/umbraco-doc-type-grid-editor/commit/74a842a1de897b8fdcf3c102edc7f0f139eec855#diff-bd0734d76903ae8b9744d8abbc39eef1L38, and with no strategy for clearing that cache, it was obvius that this bug would happen. I have changed it back to RequestCache, and it works on my machine™ :)

bjarnef commented:
@skttl I have tested the latest version and it now updates the grid content. However I wonder if the editor actually is cached - if I add a change to the partial view, it is updated right away with `debug="false"`. Shouldn't it only update it on publish?

skttl commented:
@bjarnef it is only cached per request now. That is how it was in v7 also.

ArnoldV commented:
@skttl thanks! Now you get this when no preview-layout is set:

![image](https://user-images.githubusercontent.com/4798968/58563659-d3cad200-822b-11e9-8a6e-b273c31058f6.png)

The icon, the title and the description based on @bjarnef suggestion suggestion.

skttl commented:
I think we are ready for review. @ArnoldV and I, have discussed the possibility of adding validation, but as discussed in #114, it's not completely possible. Anyway, it's out of the scope of this PR.

We added some new stuff

- New previews for non-preview editors https://github.com/umco/umbraco-doc-type-grid-editor/pull/137#issuecomment-496952459
- Editors can decide if they need a large or small dialog using the `largeDialog` config setting.
- Editors can decide if they need a grid of content types, or the default list using the `showDocTypeSelectAsGrid` config setting
- The content type select dialog now has a search field.
- Editors is using the editorService, which enables infinite editing 🎉
- Only Element content types are allowed as Doc Type Grid Editor content types.

@leekelleher @mattbrailsford anything else you would like to get in?

AndersBrohus commented:
Is there any special way to install this package? 😬 
Because i have created an test projekt, installed Umbraco 8.0.2, and downloaded the zip package from the first link at the top in the first comment and i can't make Umbraco see the DGTE grid editor..
Tried upping the ClientDependency, restarted app pool, changed modelsbuilder from purelive to dll mode, to force everything to restart so i should become a part of the solution but it simply won't show up .. :)

Inside my app_plugins i can see the DGTE and everything is there.. :)

skttl commented:
Hi @AndersBrohus 

You need to create an editor yourself. There is an example in /App_Plugins/DocTypeGridEditor/Config you can use.

AndersBrohus commented:
@skttl - I did see it but was thinking it would include it by itself :P

Thanks will try test it now :)

bjarnef commented on src/Our.Umbraco.DocTypeGridEditor/Bootstrap.cs:
```diff
@@ -42,17 +39,20 @@ protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplica
+                            Current.AppCaches.RuntimeCache.ClearByKey(
+                                string.Concat(
+                                    "Our.Umbraco.DocTypeGridEditor.Helpers.DocTypeGridEditorHelper.GetContentTypesByAlias_",
+                                    item.Alias));
```
@skttl should we replace this with `$ - string interpolation`?
```
Current.AppCaches.RuntimeCache.ClearByKey($"Our.Umbraco.DocTypeGridEditor.Helpers.DocTypeGridEditorHelper.GetContentTypesByAlias_{item.Alias}");
```

skttl replied:
Does it really matter? https://koukia.ca/string-interpolation-vs-string-format-string-concat-and-string-builder-performance-benchmarks-c1dad38032a

Besides, it shouldnøt be showstopper for the PR.

bjarnef replied:
@skttl no, it compiles to the same. I just think the newer literal notation is a bit cleaner :)

bjarnef commented:
@skttl no, it compiles to the same. I just think the newer literal notation is a bit cleaner :)

miackers commented:
Hi,

Very excited about this release - thanks for all the great work!

I tried the alpha version (0.6.2.228) and I get issues (preview don't work) when the node I am working with has not previously been published. Partial error:

Request: POST /umbraco/backoffice/DocTypeGridEditorApi/DocTypeGridEditorApi/GetPreviewMarkup?dtgePreview=1&pageId=XXX 

> "ExceptionMessage":"Object reference not set to an instance of an object.","ExceptionType":"System.NullReferenceException","StackTrace":" at Our.Umbraco.DocTypeGridEditor.Models.UnpublishedContent..ctor(IContent content, ServiceContext serviceContext)\r\n   at Our.Umbraco.DocTypeGridEditor.Web.Controllers.DocTypeGridEditorApiController.GetPreviewMarkup(PreviewData data, Int32 pageId)\r\n   at lambda_method(Closure , Object , Object[] )\r\n 

This was installed via nupkg file, and not able to debug at the moment. If I manage to find out more, I'll let you know.

skttl commented:
Hi @miackers 

Thank you very much for reporting this! I have fixed the missing methods, causing the exception :)

miackers commented:
Amazing - thanks!

rich-hamilton commented:
So glad I found this, thanks for the good work. Anyone got a simple template of how to access values in a preview?
Using IPublishedElement now?

skttl commented:
Hi @rich-hamilton

You can do like you normally do with IPublishedContent (Model.Value("alias")), or use modelsbuilder (Model.Alias)

rich-hamilton commented:
> Hi @rich-hamilton
> 
> You can do like you normally do with IPublishedContent (Model.Value("alias")), or use modelsbuilder (Model.Alias)

Thanks but what type of page do I inherit from


> Hi @rich-hamilton
> 
> You can do like you normally do with IPublishedContent (Model.Value("alias")), or use modelsbuilder (Model.Alias)

But I'm just getting this Our.Umbraco.DocTypeGridEditor.Models.DetachedPublishedElement to model type Our.Umbraco.DocTypeGridEditor.Web.PreviewModel.'

So a quick example would be great to get me up and running again.

skttl commented:
Hi Rich

Like this:

```
@inherits UmbracoViewPage<IPublishedElement>
<h1>@Model.Value("heading")</h1>
```

or using ModelsBuilder

```
@inherits UmbracoViewPage<MyHeadingType>
<h1>@Model.Heading</h1>
```

rich-hamilton commented:
> Hi Rich
> 
> Like this:
> 
> ```
> @inherits UmbracoViewPage<IPublishedElement>
> <h1>@Model.Value("heading")</h1>
> ```
> 
> or using ModelsBuilder
> 
> ```
> @inherits UmbracoViewPage<MyHeadingType>
> <h1>@Model.Heading</h1>
> ```

Thanks @skttl - perfect.

cleversolutions commented:
I don't know if this is a DTGE thing in general, or specific to the V8 port. Umbraco.AssignedContentItem throws the exception "Cannot return the IPublishedContent because the UmbracoHelper was not constructed with an IPublishedContent." when called by DocTypeGridEditorPreviewer. cshtml

To work around this, I've added to  DocTypeGridEditorPreviewer. cshtml ```TempData["CurrentPage"] = Model.Page;``` and DocTypeGridEditor.cshtml ```TempData["CurrentPage"] = Umbraco.AssignedContentItem;``` so I can access the current page in my views with ```TempData["CurrentPage"] as IPublishedContent;``` It seems kludgy to be updating views in App_Plugins. Is there a supported way of getting the IPublishedContent that works in both preview and view?

Thanks

ArnoldV commented:
> I don't know if this is a DTGE thing in general, or specific to the V8 port. Umbraco.AssignedContentItem throws the exception "Cannot return the IPublishedContent because the UmbracoHelper was not constructed with an IPublishedContent." when called by DocTypeGridEditorPreviewer. cshtml
> 
> To work around this, I've added to DocTypeGridEditorPreviewer. cshtml `TempData["CurrentPage"] = Model.Page;` and DocTypeGridEditor.cshtml `TempData["CurrentPage"] = Umbraco.AssignedContentItem;` so I can access the current page in my views with `TempData["CurrentPage"] as IPublishedContent;` It seems kludgy to be updating views in App_Plugins. Is there a supported way of getting the IPublishedContent that works in both preview and view?
> 
> Thanks

DTGE in v8 works with "PublishedElement" instead of PublishedContent, since it's not a content page, but an element on a page. What do you want to achieve, there might be a different solution to this.

ArnoldV commented:
@leekelleher @mattbrailsford is there something you want us to do to keep this moving forward to an 'official' release? Or is it, at this point, mostly a 'having time available to check' issue?

leekelleher commented:
@ArnoldV I'll have a chat with @mattbrailsford on how we should proceed. 👍

cleversolutions commented:
@ArnoldV I'm trying to get a reference to the IPublishedContent that contains the PublishedElement. Specifically, I have an editor with DTGE that renders navigation components. By default it will render navigation for all of the child nodes of the current page, but to do so I need to get a reference to the current page.

As an aside, DTGE in V8 is working really well. I'm leaning very heavily on it for a new website we hope to have live by end of year. Support for the large dialog was a nice add.

cleversolutions commented:
@skttl I added GetContainerPage to HtmlHelperExtensions so developers can access the containing page in https://github.com/cleversolutions/umbraco-doc-type-grid-editor/commit/cb59ec2084ea1563bf221f00d27623969f09cff1. Is this something you would like to merge into your PR?

skttl commented:
@cleversolutions I appreciate your effort, and I see the need for a simpler way of getting the current page. But I don't think it is a responsibility of DTGE do do that, as the problem also exists in NC and similar packages.

bjarnef commented:
@cleversolutions you can always use one of the following to get the current node in the partial view:

```
var currentNode = UmbracoContext.Current.PublishedContentRequest.PublishedContent;
var root = currentNode.Root();
```

or 

```
var currentNode = Umbraco.AssignedContentItem.Root();
var root = currentNode.Root();
```

https://our.umbraco.com/forum/templates-partial-views-and-macros/94410-get-current-root-node-from-nested-content-partial-view

cleversolutions commented:
Thanks @bjarnef, that works great for front end requests; however when generating previews, both Umbraco.AssignedContentItem and UmbracoContext.PublishedRequest are null. @skttl is right, this is an issue that affects other packages, which is why LeBlender added a similar extension method. Here is another alternative idea, the config currently allows you to change the renderer, but not the previewer view. If we allow people to use their own previewer, they can then do whatever terrible smelly they want with the PreviewModel.

GertyEngrie commented:
I this actually merge-ready or is there still work involved @leekelleher? Happy to help out :-)

leekelleher commented:
@GertyEngrie I haven't been able to chat with Matt about this yet to decide how we'd like to proceed. This mostly comes down to how DTGE v8 will be supported going forwards.

Personally (not speaking for Matt), I'm less interested in actively supporting a package that I'm not actively using. That said, I respect that it is actively used and popular... so, once Matt and I have agreed a way forward, we'll chat with Søren, for his involvement/approval.

bjarnef commented:
@skttl I just tested upgrading Umbraco v8.0.2 til v8.1.0 and get the following error on startup.

The `IPublishedContent` has been refactored, so it seems some of the (extension)methods has changed.

![image](https://user-images.githubusercontent.com/2919859/60874568-8aad6b00-a238-11e9-9f2f-7d9f7d546e53.png)

cleversolutions commented:
I've updated the package for 8.1 in commit https://github.com/cleversolutions/umbraco-doc-type-grid-editor/commit/442526b407c26cab79d9b4eb665815149fd507b2 I've done some minimal testing with it and so far it seems okay. I'll submit a pull request to @skttl's fork.

bjarnef commented:
@skttl I just tried the lastest build which fix the startup error, but I get these console errors for the preview in backoffice.

![image](https://user-images.githubusercontent.com/2919859/60946102-9f980600-a2ed-11e9-9b5e-8ca709bce20b.png)

```
@inherits Umbraco.Web.Mvc.UmbracoViewPage<ContentModels.NestedBox>
@using ContentModels = Umbraco.Web.PublishedModels;
@{
    var site = Umbraco.AssignedContentItem.Root();
    <strong>@($"{site.Id} {site.Name}")</strong>
}
<div>
    @if (Model.Image != null)
    {
        <img src="@Model.Image.GetCropUrl(200, 200)" alt="" />
    }
    <h3>@Model.Heading</h3>
    <p>@Model.Text</p>
</div>
```

skttl commented:
Thanks for testing @bjarnef.

I think this comes from your use of AssignedContentItem, which IMO is invalid when used in an IPublishedElement. You won't be able to get root from that either.

AssignedContentItem is the same as the content used to initiate the UmbracoHelper, which in this case is the IPublishedElement of the editor.

To get your root, you have to use the UmbracoContext like `UmbracoContext.PublishedRequest.PublishedContent.Root()`

bjarnef commented:
Strange because it was also `IPublishedElement` before upgrading Umbraco to v8.1.0 and in frontend it does output node name and node id of the root node fine :)

skttl commented:
Ah, must be something in the previewer that is different from Umbraco behavior then :)

bjarnef commented:
@skttl I tried with `UmbracoContext.PublishedRequest.PublishedContent.Root()` and it is the same issue as when using `Umbraco.AssignedContentItem.Root()`

skttl commented:
@bjarnef is your node published?

bjarnef commented:
@skttl of course, it is the site node (homepage) and on the same page the grid editor is inserted in grid :)

skttl commented:
weird - I'll look into it.

Can you do other things with the Umbraco helper? or does that break too?

bjarnef commented:
I works if I request the node direct by id e.g. `Umbraco.Content(1234)`.

Both of these works in frontend, but throw console errors in backoffice.
`UmbracoContext.PublishedRequest.PublishedContent.Root()`
`Umbraco.AssignedContentItem.Root()`

bjarnef commented:
When using the following:

```
var site = UmbracoContext.PublishedRequest.PublishedContent;
<strong>@($"{site.Id} {site.Name}")</strong>
```

![image](https://user-images.githubusercontent.com/2919859/60973192-105c1400-a328-11e9-9c3d-6623b1582ede.png)

I guess it can't find current page in this case or hasn't a pageId.
https://github.com/umco/umbraco-doc-type-grid-editor/blob/9a13c1ae301a90eaf18592fe94797effc334c774/src/Our.Umbraco.DocTypeGridEditor/Web/Controllers/DocTypeGridEditorApiController.cs#L144

bjarnef commented:
@skttl it seems both examples work in frontend, when rendered in the grid. It seems to be an issue when used in the preview mechanism.

skttl commented:
Just tried my old demo solution (8.0.2) and it does the same when using `Umbraco.AssignedContentItem`

skttl commented:
Turns out we did need this hack :)
https://github.com/umco/umbraco-doc-type-grid-editor/blob/74a842a1de897b8fdcf3c102edc7f0f139eec855/src/Our.Umbraco.DocTypeGridEditor/Web/Controllers/DocTypeGridEditorApiController.cs#L161

bjarnef commented:
@skttl it seems to work great now in the preview as well. It output the node name and node id.

![image](https://user-images.githubusercontent.com/2919859/61037959-65e9fc80-a3cc-11e9-8fb6-2916c25d1b65.png)

GeorgePhillipson commented:
Downloaded Our.Umbraco.DocTypeGridEditor.0.6.2-alpha-000231.nupkg and installed last night (14 July 2019)
I'm using:
Umbraco 8.1.0
Umbraco.ModelsBuilder 8.1.0
Umbraco.ModelsBuilder.API 8.1.0

Tested using Nested Content, Images etc and DocTypeGridEditor works OK, no issues found

cmwalolo commented:
When using the grid without MVCRenderController  for the document types => all is fine.
When I go through a controller I get this error, the passed model seems to be the current page model and not the specific grid model : 

Cannot bind source type Crolow.Core.Models.Umbraco.Home to model type Crolow.Core.Models.Umbraco.GridSingleCard.

bjarnef commented:
@skttl it seems there is an issue with previewing grid editor in backoffice if `previewViewPath` for the grid editor config in either `/config/grid.editors.config.js/` or from a `package.manifest` in a folder in `/App_Plugins` is not inside `/Views/` folder.

I think it is because it look for a view or partial view in the view engines, which razor view engine have some default folders, where is look for views.

https://github.com/umco/umbraco-doc-type-grid-editor/blob/3b3a8923a016cd3c63f9796d3707fcb8db97d17d/src/Our.Umbraco.DocTypeGridEditor/Web/Helpers/ViewHelper.cs#L52-L62

But is seems to be possible to configurate razor view engine to use custom locations.
https://marisks.net/2017/02/03/razor-view-engine-for-feature-folders/
https://www.ryadel.com/en/asp-net-mvc-add-custom-locations-to-the-view-engine-default-search-patterns/


This works:

```
{
    "gridEditors": [
        {
            "name": "Nested Box",
            "alias": "nestedBox",
            "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
            "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
            "icon": "icon-brick",
            "config": {
                "allowedDocTypes": [
                  "nestedBox"
                ],
                "nameTemplate": "",
                "enablePreview": true,
                "viewPath": "/App_Plugins/NestedBoxGridEditor/Views/",
                "previewViewPath": "/Views/Partials/",
                "previewCssFilePath": "",
                "previewJsFilePath": ""
              }
        }
    ]
}
```

or this:

```
{
    "gridEditors": [
        {
            "name": "Nested Box",
            "alias": "nestedBox",
            "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
            "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
            "icon": "icon-brick",
            "config": {
                "allowedDocTypes": [
                  "nestedBox"
                ],
                "nameTemplate": "",
                "enablePreview": true,
                "viewPath": "/Views/Partials/",
                "previewViewPath": "/Views/Partials/",
                "previewCssFilePath": "",
                "previewJsFilePath": ""
              }
        }
    ]
}
```

This doesn't work:
```
{
    "gridEditors": [
        {
            "name": "Nested Box",
            "alias": "nestedBox",
            "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
            "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
            "icon": "icon-brick",
            "config": {
                "allowedDocTypes": [
                  "nestedBox"
                ],
                "nameTemplate": "",
                "enablePreview": true,
                "viewPath": "/Views/Partials/",
                "previewViewPath": "/Views/Partials/",
                "previewCssFilePath": "",
                "previewJsFilePath": ""
              }
        }
    ]
}
```



This doesn't work:

```
{
    "gridEditors": [
        {
            "name": "Nested Box",
            "alias": "nestedBox",
            "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
            "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
            "icon": "icon-brick",
            "config": {
                "allowedDocTypes": [
                  "nestedBox"
                ],
                "nameTemplate": "",
                "enablePreview": true,
                "viewPath": "/App_Plugins/NestedBoxGridEditor/Views/",
                "previewViewPath": "/App_Plugins/NestedBoxGridEditor/Views/",
                "previewCssFilePath": "",
                "previewJsFilePath": ""
              }
        }
    ]
}
```

![image](https://user-images.githubusercontent.com/2919859/62322869-567b4200-b4a6-11e9-8d76-9936e422022f.png)

skttl commented:
Hi @bjarnef, that error message looks more like you have a null reference on your Image property.

bjarnef commented:
@skttl no, it isn't :)

![image](https://user-images.githubusercontent.com/2919859/62355191-d0dfac80-b50e-11e9-8f10-aedf889c5179.png)

It works fine when specifying `/Views/Partials/` as `viewPath` and `previewViewPath`, but if path to one of these is `/App_Plugins/NestedBoxGridEditor/Views/` it returns this console error.

I guess it is more an issue when rendering the preview, because it looks for a partial view and I guess views in `/App_Plugins` by default not is one of the path razor engine search for partial views.

bjarnef commented:
@skttl okay, it turned out that the preview also broke, with this configuration, if an component like this is added to the project. https://our.umbraco.com/forum/using-umbraco-and-getting-started/96824-custom-apicontroller-in-umbraco-8-cannot-be-reached#comment-306194

```
{
  "gridEditors": [
    {
      "name": "Nested Box",
      "alias": "nestedBox",
      "view": "/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html",
      "render": "/App_Plugins/DocTypeGridEditor/Render/DocTypeGridEditor.cshtml",
      "icon": "icon-brick",
      "config": {
        "allowedDocTypes": [
          "nestedBox"
        ],
        "nameTemplate": "",
        "enablePreview": true,
        "viewPath": "/Views/Partials/",
        "previewViewPath": "/Views/Partials/",
        "previewCssFilePath": "",
        "previewJsFilePath": ""
      }
    }
  ]
}
```

when I comment this out the preview work with the example above.

```
[RuntimeLevel(MinLevel = RuntimeLevel.Run)]
internal class ApiConfigurationComposer : IComposer
{
    public void Compose(Composition composition)
    {
        //composition.Components().Append<ApiConfigurationComponent>();
    }

    public class ApiConfigurationComponent : IComponent
    {
        public void Initialize()
        {
            RouteTable.Routes.MapMvcAttributeRoutes();
            GlobalConfiguration.Configuration.MapHttpAttributeRoutes();
            GlobalConfiguration.Configuration.Formatters.Clear();
            GlobalConfiguration.Configuration.Formatters.Add(new JsonMediaTypeFormatter
            {
                SerializerSettings = new JsonSerializerSettings
                {
                    ContractResolver = new CamelCasePropertyNamesContractResolver(),
                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                    Formatting = Formatting.Indented
                }
            });
        }

        public void Terminate()
        {
        }
    }
}
```

![image](https://user-images.githubusercontent.com/2919859/62372739-8d019d00-b538-11e9-9aa2-16e7f1dc03e8.png)

bjarnef commented:
@skttl I found this old issue for Umbraco: https://issues.umbraco.org/issue/U4-8614 and with this change https://github.com/umbraco/Umbraco-CMS/commit/bc2b9ab29850e6473f7c7b6724a5f0e9e1e7a563

Using this approach instead, it seems DTGE preview still works.

I guess that `.Clear()` or `.Add(new JsonMediaTypeFormatter { })` here conflicted with the implementation of DTGE.

miackers commented:
Hey guys - really great work with this port!! I've been using it for more than a month now and no problems. However, I did just notice that it looks like macros inside RTE inside a grid is not rendering - seem to recall a same problem with this on V7 a few years back.

Managed to get around it using similar code I used back then, but just checking if anyone else is seeing the same problem (and perhaps i am just doing something wrong?)

`
public static string RenderContentMacro(this UmbracoHelper UmbracoHelper, string content)
        {
            if (string.IsNullOrWhiteSpace(content))
                return string.Empty;

            if (content.IndexOf("UMBRACO_MACRO", StringComparison.InvariantCultureIgnoreCase) == -1)
                return content;

            var html = new HtmlDocument();
            html.LoadHtml(content);

            var macroNodes = html.DocumentNode.SelectNodes("//comment()[contains(., '<?UMBRACO_MACRO')]");

            if (macroNodes == null || !macroNodes.Any())
                return content;

            foreach (var c in macroNodes)
            {
                var macroMatch = MACRO_FORMAT.Match(c.InnerHtml);

                if (macroMatch.Value.IndexOf("macroalias", StringComparison.InvariantCultureIgnoreCase) == -1)
                    continue;

                var attributes = GetAttributesFromElement(macroMatch.Value.ToString());

                content = content.Replace(c.ParentNode.OuterHtml, UmbracoHelper.RenderMacro(attributes["macroalias"].ToString(), attributes).ToString());
            }

            return content;
        }

        internal static Dictionary<string, object> GetAttributesFromElement(string tag)
        {
            var m =
                Regex.Matches(tag, "(?<attributeName>\\S*)=\"(?<attributeValue>[^\"]*)\"",
                              RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace);
            // fix for issue 14862: return lowercase attributes for case insensitive matching
            var d = m.Cast<Match>().ToDictionary(attributeSet => attributeSet.Groups["attributeName"].Value.ToString().ToLower(), attributeSet => (object)attributeSet.Groups["attributeValue"].Value.ToString());
            return d;
        }
`
